File: Ds/Stack/Stack_cs/Stack.cs
using System;

//LIFO
//A5er wa7ed da5al howa awel wa7ed ye5rog
public class Stack<T>
{
	private class Node
	{
		public T item;
		public Node next;
	}
	Node stack;
	int curIndex = 0;
	public Stack ()
	{
		stack = new Node();
		stack.next = null;
	}
	public void Push(T item)		//INSERT an item into the stack
	{
		Node temp = new Node ();
		temp.item = item;
		temp.next = stack;
		stack = temp;
		curIndex++;
	}
	public T Pop()				//removes Item from stack
	{
		if (curIndex == 0)
			return default(T);
		T item = stack.item;
		stack = stack.next;
		curIndex--;
		System.GC.Collect ();
		return item;
	}
	public bool isEmpty()
	{
		return curIndex == 0;
	}
	public int Size()
	{
		return curIndex;
	}
	public void Print()
	{
		Node it = stack;
		while(it != null)
		{
			Console.WriteLine ( it.item );
			it = it.next;
		}
	}
}

File: Ds/Stack/Stack_cpp/Stack.cpp
#include "Stack.h"

template<class T>
Stack<T>::Stack()
{
    tail = NULL;
    sz = 0;
}

template<class T>
Stack<T>::Stack(T value, int intial_size)
{
    sz = intial_size;
    tail = NULL;
    while(intial_size > 0)
    {
        Node* temp = new Node(value, tail);
        tail = temp;
        intial_size--;
    }
}

template<class T>
T& Stack<T>::Top()
{
    try{
        if(tail != NULL)
            return tail->value;
        throw(0);
    }catch(...){cout << "EmptyStack" << endl;}
}

template<class T>
void Stack<T>::Pop()
{
    sz--;
    if(tail == NULL)
        return;
    Node* temp = tail;
    tail = tail->next;
    delete temp;
}
template<class T>
void Stack<T>::Push(T value)
{
    Node* temp = new Node(value, tail);
    tail = temp;
    sz++;
}
template<class T>
int Stack<T>::Size()
{
    return sz;
}

template<class T>
void Stack<T>::Print()
{
    Node* temp = tail;
    while(temp != NULL)
    {
        cout << temp->value << " ";
        temp = temp->next;
    }
}


template<class T>
Stack<T>::~Stack()
{
    Node* temp = tail;
    while(temp != NULL)
    {
        temp = tail;
        tail = tail->next;
        delete temp;
    }
}

template class Stack<int>;


